### PR TITLE
Feature/rdkb code porting

### PR DESCRIPTION
**Pull request by morantr**
Port the latest changes from RDKB.

### CODE REVIEW

**Review by tomereli — DISMISSED** [review]
Comments inline

**Comment by tomereli** [comment]
[port rdkb chnages to prplMesh](https://trello.com/c/i3wVVvpN/25-port-rdkb-chnages-to-prplmesh)

File: CMakeLists.txt [hunk 1]
```diff
@@ -60,6 +60,7 @@ set(BEEROCKS_BH_WIRE_IFACE "eth1")
 set(BEEROCKS_STOP_ON_FAILURE_ATTEMPTS 1)
 set(BEEROCKS_NOTIFY 0)
 set(BEEROCKS_REPEATER_MODE 0)
+set(BEEROCKS_LOG_PATH "${TMP_PATH}/logs")
```
Comment (arnout) [inline CMakeLists.txt]: I want to evolve to a situation where TARGET_PLATFORM is no longer used, and instead we have cache variables that are set by the platform integration (i.e. openwrt Makefile, yocto recipe, or toolchainfile). That is IMO a cleaner way to manage things. So, I would make this a CACHE variable.

Nitpick, because it can easily be done together with moving all the other variations to cache variables.

File: agent/src/beerocks/monitor/beerocks_monitor_main.cpp [hunk 2]
```diff
@@ -165,8 +165,8 @@ int main(int argc, char *argv[])
 
     //redirect stdout / stderr to file
     // int fd_log_file_std = beerocks::os_utils::redirect_console_std("/dev/null");
-    int fd_log_file_std = beerocks::os_utils::redirect_console_std(beerocks_slave_conf.sLog.path +
-                                                                   base_monitor_name + "_std.log");
+    int fd_log_file_std = beerocks::os_utils::redirect_console_std(
+        beerocks_slave_conf.sLog.path + base_monitor_name + "_std.txt.0");
```
Comment (arnout) [inline agent/src/beerocks/monitor/beerocks_monitor_main.cpp]: Why the change from `.log` to `.txt.0`?

Since this is such a big (but largely mechanical) change, it should be in a separate commit.

File: common/beerocks/bcl/source/beerocks_logging.cpp [hunk 3]
```diff
@@ -400,7 +400,7 @@ std::string logging::get_log_filename()
     if (m_log_filename.size() > 0) {
         return m_log_filename;
     }
-    return m_module_name + ".%datetime{%Y%M%d_%H%m%s}.log";
+    return m_module_name + ".txt.0";
```
Comment (arnout) [inline common/beerocks/bcl/source/beerocks_logging.cpp]: Why remove the datetime (and overwrite the old log)?

The commit message says that you do it, but not why. The why is the important part.

File: common/beerocks/scripts/beerocks_utils.sh [hunk 4]
```diff
@@ -24,8 +24,9 @@ fi
 
 BEEROCKS_TMP_PATH=/tmp/beerocks
 
-BEEROCKS_LOGS_PATH=$BEEROCKS_TMP_PATH/logs
-BEEROCKS_ERRORS_FILE=$BEEROCKS_TMP_PATH/beerocks_errors.log
+BEEROCKS_LOGS_PATH=/rdklogs/logs
+
+BEEROCKS_ERRORS_FILE=$BEEROCKS_LOGS_PATH/beerocks_errors.txt.0
```
Comment (arnout) [inline common/beerocks/scripts/beerocks_utils.sh]: I think it would be better to refactor this script with prplmesh_utils.sh and keep only one of them. That is, assuming that this script is actually still used on rdkb.

File: framework/CMakeLists.txt [hunk 5]
```diff
@@ -42,6 +42,10 @@ option(BUILD_TESTS "build multiap unit tests" OFF)
 set(MSGLIB "zmq" CACHE STRING "Which messaging library backend to use")
 set_property(CACHE MSGLIB PROPERTY STRINGS "zmq" "nng" "None")
 
+if (TARGET_PLATFORM STREQUAL "rdkb")
+	set(MSGLIB "None")
```
Comment (arnout) [inline framework/CMakeLists.txt]: As already said by @tomereli : NACK on this. Just pass -DMSGLIB=None while building.

File: controller/src/beerocks/master/tasks/rdkb/rdkb_wlan_task.cpp [hunk 6]
```diff
@@ -266,10 +265,11 @@ void rdkb_wlan_task::handle_event(int event_type, void *obj)
         if (obj) {
             auto event_obj         = (steering_client_disconnect_request_event *)obj;
             std::string client_mac = net::network_utils::mac_to_string(event_obj->client_mac);
-            TASK_LOG(INFO) << "STEERING_CLIENT_DISCONNECT_REQUEST event was received client_mac "
-                           << client_mac;
+            TASK_LOG(INFO) << "STEERING_CLIENT_DISCONNECT_REQUEST received for " << client_mac;
             son_actions::disconnect_client(database, cmdu_tx, client_mac, event_obj->bssid,
                                            event_obj->type, event_obj->reason);
+
+            add_pending_events(int(STEERING_CLIENT_DISCONNECT_RESPONSE), event_obj->sd);
```
Comment (arnout) [inline controller/src/beerocks/master/tasks/rdkb/rdkb_wlan_task.cpp:272]: The commit message for this change was not really understandable for me. Suggested improvement:

```
controller: rdkb_wlan_task: properly respond to STEERING_CLIENT_DISCONNECT

No response was sent to the STEERING_CLIENT_DISCONNECT_REQUEST,
which caused bml_rdkb_steering_client_disconnect to always report failure.

Queue a response after starting the disconnect_client action.

Also modify some prints.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
```

File: common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp [hunk 7]
```diff
@@ -567,7 +567,7 @@ bool ap_wlan_hal_dwpal::sta_disassoc(int8_t vap_id, const std::string &mac, uint
     // Build command string
     const std::string ifname =
         beerocks::utils::get_iface_string_from_iface_vap_ids(m_radio_info.iface_name, vap_id);
-    std::string cmd = "DISASSOCIATE " + ifname + " " + mac + " tx=0";
+    std::string cmd = "DISASSOCIATE " + ifname + " " + mac;
```
Comment (arnout) [inline common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:581]: Again, I didn't really understand the commit message. Is the problem that tx=0 is not understood by hostapd? And that this is detected by sniffing the wireless and observing that the disassociate frame doesn't come? Or does tx=0 actually mean to disassociate without sending a disassociate frame?

File: common/beerocks/bcl/include/bcl/son/son_wireless_utils.h [hunk 8]
```diff
@@ -140,7 +147,7 @@ class wireless_utils {
 
     // clang-format off
     static constexpr sPhyRateTableEntry phy_rate_table[PHY_RATE_TABLE_ANT_MODE_MAX][PHY_RATE_TABLE_MCS_MAX] = {
-            // 1X1_SS1_table
+        // 1X1_SS1_table
```
Comment (arnout) [inline common/beerocks/bcl/include/bcl/son/son_wireless_utils.h:150]: Should have been part of previous commit.

File: common/beerocks/bcl/include/bcl/son/son_wireless_utils.h [hunk 9]
```diff
@@ -152,8 +159,8 @@ class wireless_utils {
         {16, 13,    {{650,  722, -673},     {1350,    1500,   -647},	{2925,  3250,   -632}, {5850,  6500,   -602}}},
         {0,  10,    {{780,  867, -644},     {1620,    1800,   -619},	{3510,  3900,   -587}, {7020,  7800,   -557}}},
         {0,  8,     {{0,    0,   0},        {1800,    2000,   -599},	{3900,  4333,   -558}, {7800,  8667,   -528}}},
-            },
-            // 2X2_SS1_table
+        },
+        // 2X2_SS1_table
```
Comment (arnout) [inline common/beerocks/bcl/include/bcl/son/son_wireless_utils.h:163]: Should have been part of previous commit.

File: common/beerocks/bcl/include/bcl/son/son_wireless_utils.h [hunk 10]
```diff
@@ -165,8 +172,8 @@ class wireless_utils {
         {16, 13,    {{650, 722, -735},      {1350,    1500,   -709},    {2925,    3250,   -684}, {5850,  6500,   -654}}},
         {0,  10,    {{780, 867, -699},      {1620,    1800,   -674},    {3510,    3900,   -637}, {7020,  7800,   -607}}},
         {0,  8,     {{0,   0,   0},         {1800,    2000,   -659},    {3900,    4333,   -614}, {7800,  8667,   -584}}},
-            },
-            // 2X2_SS2_table
+        },
+        // 2X2_SS2_table
```
Comment (arnout) [inline common/beerocks/bcl/include/bcl/son/son_wireless_utils.h:176]: Should have been part of previous commit.

File: common/beerocks/bcl/include/bcl/son/son_wireless_utils.h [hunk 11]
```diff
@@ -181,7 +188,7 @@ class wireless_utils {
         }
     };
 
-    static constexpr sPhyRateBitRateEntry bit_rate_max_table_mbps[BIT_RATE_MAX_TABLE_SIZE] = {
+    static constexpr sPhyRateBitRateEntry bit_rate_max_table_mbps[BIT_RATE_MAX_TABLE_SIZE] = {
```
Comment (arnout) [inline common/beerocks/bcl/include/bcl/son/son_wireless_utils.h]: This introduces a whitespace error.

File: common/beerocks/bcl/source/son/son_wireless_utils.cpp [hunk 12]
```diff
@@ -32,20 +32,21 @@ wireless_utils::estimate_ul_params(int ul_rssi, uint16_t sta_phy_tx_rate_100kb,
...
-
-    int max_ant_mode = (sta_capabilities->ant_num == beerocks::ANT_1X1)
-                           ? beerocks::ANT_MODE_1X1_SS1
-                           : beerocks::ANT_MODE_2X2_SS2;
-    int max_mcs = (is_5ghz && (sta_capabilities->wifi_standard & int(beerocks::STANDARD_AC)))
-                      ? sta_capabilities->vht_mcs
-                      : sta_capabilities->ht_mcs;
+    int estimated_ul_rssi_lut_delta_min          = 120 * 10;
+    static const sPhyUlParams estimation_minimal = {
+        is_5ghz ? phy_rate_table[0][0].tx_power_5 : phy_rate_table[0][0].tx_power_2_4,
```
Comment (arnout) [inline common/beerocks/bcl/source/son/son_wireless_utils.cpp]: Using a function argument (is_5ghz) to initialize a static variable is not a good idea... In other words, it shouldn't be static.

File: common/beerocks/bcl/source/son/son_wireless_utils.cpp [hunk 13]
```diff
@@ -32,20 +32,21 @@ wireless_utils::estimate_ul_params(int ul_rssi, uint16_t sta_phy_tx_rate_100kb,
...
+                                                    ESTIMATION_FAILURE_INVALID_RSSI};
+    sPhyUlParams estimation_normal               = estimation_invalid;
+
+    static const int max_ant_mode = (sta_capabilities->ant_num == beerocks::ANT_1X1)
+                                        ? beerocks::ANT_MODE_1X1_SS1
+                                        : beerocks::ANT_MODE_2X2_SS2;
+    static const int max_mcs =
+        (is_5ghz && (sta_capabilities->wifi_standard & int(beerocks::STANDARD_AC)))
+            ? sta_capabilities->vht_mcs
+            : sta_capabilities->ht_mcs;
```
Comment (arnout) [inline common/beerocks/bcl/source/son/son_wireless_utils.cpp]: Again, these can't be static.

File: common/beerocks/bcl/source/son/son_wireless_utils.cpp [hunk 14]
```diff
@@ -55,43 +56,110 @@ wireless_utils::estimate_ul_params(int ul_rssi, uint16_t sta_phy_tx_rate_100kb,
...
-        for (int bw = max_bw; bw > -1; bw--) {                     // filter by mac_bw
+    LOG(DEBUG) << " | UL RSSI:" << ul_rssi << " | sta_phy_tx_rate:" << sta_phy_tx_rate_100kb / 10
+               << " Mbps | AP BW:"
+               << beerocks::utils::convert_bandwidth_to_int((beerocks::eWiFiBandwidth)ap_bw)
+               << " | is_5ghz:" << is_5ghz << " | ant_num:" << int(sta_capabilities->ant_num)
+               << " | max_ant_mode:" << max_ant_mode << " | max MCS:" << max_mcs << " | max BW:"
+               << beerocks::utils::convert_bandwidth_to_int((beerocks::eWiFiBandwidth)max_bw);
+
+    if (ul_rssi == beerocks::RSSI_INVALID) {
+        LOG(DEBUG) << __FUNCTION__ << " | Can not estimate UL parameters (invalid RSSI)";
```
Comment (arnout) [inline common/beerocks/bcl/source/son/son_wireless_utils.cpp]: This change in the debug output is a bit annoying because it screws up the diff, which complicates review (and also later understanding of this change, in case it needs to be revisited). However, I'm not so sure if it's a useful change.

- If you want to see the function, just set the format to `"%datetime %level [%logger] %func | %msg"` in the logging config.
- I'm not sure what the value is of the additional pipes; it's not really a table, is it? And if it is, it anyway won't be aligned because no field width is defined.

(nitpick)

File: common/beerocks/bcl/source/son/son_wireless_utils.cpp [hunk 15]
```diff
@@ -55,43 +56,110 @@ wireless_utils::estimate_ul_params(int ul_rssi, uint16_t sta_phy_tx_rate_100kb,
...
+
+    // If station phyrate value is below table's minimum, return minimal estimation
+    if (sta_phy_tx_rate_100kb < phy_rate_table[0][0].bw_values[0].gi_long_rate) {
+        LOG(DEBUG) << __FUNCTION__ << " | Can not estimate UL parameters (STA phyrate is too low)";
+        return estimation_minimal;
+    }
+
+    // If station phyrate value is above table's maximum, return maximal estimation
+    if (sta_phy_tx_rate_100kb >
+        phy_rate_table[max_ant_mode][max_mcs].bw_values[max_bw].gi_short_rate) {
```
Comment (arnout) [inline common/beerocks/bcl/source/son/son_wireless_utils.cpp:78]: It may be appropriate to put this in a variable `max_rate`. That also removes the need for the comment above (it becomes self-explanatory).

Same for the minimum, above.

File: common/beerocks/bcl/source/son/son_wireless_utils.cpp [hunk 16]
```diff
@@ -55,43 +56,110 @@ wireless_utils::estimate_ul_params(int ul_rssi, uint16_t sta_phy_tx_rate_100kb,
...
+                estimated_ul_rssi_lut       = phy_rate_table[ant_mode][mcs].bw_values[bw].rssi;
+                estimated_ul_rssi_lut_delta = std::abs(ul_rssi_lut - estimated_ul_rssi_lut);
 
-                // same rate && min rssi_delta
-                if ((sta_phy_tx_rate_100kb ==
-                     phy_rate_table[ant_mode][mcs].bw_values[bw].gi_long_rate) ||
-                    (sta_phy_tx_rate_100kb ==
-                     phy_rate_table[ant_mode][mcs].bw_values[bw].gi_short_rate)) {
+                auto gi_long_rate  = phy_rate_table[ant_mode][mcs].bw_values[bw].gi_long_rate;
+                auto gi_short_rate = phy_rate_table[ant_mode][mcs].bw_values[bw].gi_short_rate;
```
Comment (arnout) [inline common/beerocks/bcl/source/son/son_wireless_utils.cpp:104]: To make the code more readable, it may be more useful to instead define `auto values = phy_rate_table[ant_mode][mcs].bw_values[bw];`. That makes all of the code below much shorter.

File: common/beerocks/bcl/source/son/son_wireless_utils.cpp [hunk 17]
```diff
@@ -55,43 +56,110 @@ wireless_utils::estimate_ul_params(int ul_rssi, uint16_t sta_phy_tx_rate_100kb,
...
                     if (estimated_ul_rssi_lut_delta <= estimated_ul_rssi_lut_delta_min) {
                         estimated_ul_rssi_lut_delta_min = estimated_ul_rssi_lut_delta;
-                        estimation.tx_power = is_5ghz ? phy_rate_table[ant_mode][mcs].tx_power_5
-                                                      : phy_rate_table[ant_mode][mcs].tx_power_2_4;
-                        estimation.rssi =
-                            int(float(estimated_ul_rssi_lut) / 10.0 + 0.5); //round value
+                        estimation_normal.tx_power =
+                            is_5ghz ? phy_rate_table[ant_mode][mcs].tx_power_5
+                                    : phy_rate_table[ant_mode][mcs].tx_power_2_4;
+                        estimation_normal.rssi = int(ceil(estimated_ul_rssi_lut / 10.0));
```
Comment (arnout) [inline common/beerocks/bcl/source/son/son_wireless_utils.cpp]: Why ceil instead of approximation rounding like it was before?

File: common/beerocks/bcl/source/son/son_wireless_utils.cpp [hunk 18]
```diff
@@ -55,43 +56,110 @@ wireless_utils::estimate_ul_params(int ul_rssi, uint16_t sta_phy_tx_rate_100kb,
...
+
+    if (ul_rssi == beerocks::RSSI_INVALID) {
+        LOG(DEBUG) << __FUNCTION__ << " | Can not estimate UL parameters (invalid RSSI)";
+        return estimation_invalid;
+    }
+
+    // If station phyrate value is below table's minimum, return minimal estimation
+    if (sta_phy_tx_rate_100kb < phy_rate_table[0][0].bw_values[0].gi_long_rate) {
+        LOG(DEBUG) << __FUNCTION__ << " | Can not estimate UL parameters (STA phyrate is too low)";
+        return estimation_minimal;
```
Comment (arnout) [inline common/beerocks/bcl/source/son/son_wireless_utils.cpp]: Since this is the only place where estimation_minimal is used, I don't think it's worth defining a variable for it. Or at least, move it here. But actually, I'd just set the three fields of estimation_normal here. And then the rename of estimation_normal to estimation is unnecessary as well.

File: common/beerocks/bcl/source/son/son_wireless_utils.cpp [hunk 19]
```diff
@@ -55,43 +56,110 @@ wireless_utils::estimate_ul_params(int ul_rssi, uint16_t sta_phy_tx_rate_100kb,
...
+    LOG(DEBUG) << " | UL RSSI:" << ul_rssi << " | sta_phy_tx_rate:" << sta_phy_tx_rate_100kb / 10
+               << " Mbps | AP BW:"
+               << beerocks::utils::convert_bandwidth_to_int((beerocks::eWiFiBandwidth)ap_bw)
+               << " | is_5ghz:" << is_5ghz << " | ant_num:" << int(sta_capabilities->ant_num)
+               << " | max_ant_mode:" << max_ant_mode << " | max MCS:" << max_mcs << " | max BW:"
+               << beerocks::utils::convert_bandwidth_to_int((beerocks::eWiFiBandwidth)max_bw);
+
+    if (ul_rssi == beerocks::RSSI_INVALID) {
+        LOG(DEBUG) << __FUNCTION__ << " | Can not estimate UL parameters (invalid RSSI)";
+        return estimation_invalid;
```
Comment (arnout) [inline common/beerocks/bcl/source/son/son_wireless_utils.cpp]: I don't really think adding the estimation_invalid variable adds a lot of value. The way it was done originally works just as well and is equally clear.

File: controller/src/beerocks/master/tasks/optimal_path_task.cpp [hunk 20]
```diff
@@ -912,6 +913,7 @@ void optimal_path_task::work()
             auto hostap_sibling = it.second;
 
             int hostap_channel    = database.get_node_channel(hostap);
+            skip_estimation       = false; // initialise for each HostAP candidate
```
Comment (arnout) [inline controller/src/beerocks/master/tasks/optimal_path_task.cpp]: Then it should also be declared in this scope. So `bool skip_estimation = false;` here.

**Review by arnout — CHANGES_REQUESTED** [review]
Reviewed up to and including "controller: Add/Modify debug logs and add comments to optimal_path_task".

Most of the commits have excellent commit messages, and also the way it's split up is generally very clean. Kudos!

There's one thing that I believe is an actual bug: making some of the local variables static while they are calculated based on parameters. The static means that they are only calculated the first time.

File: CMakeLists.txt (hunk 1 above)
Comment (morantr) [reply CMakeLists.txt]: will be done as part of a different task.

File: agent/src/beerocks/monitor/beerocks_monitor_main.cpp (hunk 2 above)
Comment (morantr) [reply agent/src/beerocks/monitor/beerocks_monitor_main.cpp]: I really really don't know why they did it. There was no explanation for it in the code I ported.
@tomereli can do you know the reason.

File: common/beerocks/bcl/source/beerocks_logging.cpp (hunk 3 above)
Comment (morantr) [reply common/beerocks/bcl/source/beerocks_logging.cpp]: Same answer.
https://github.com/prplfoundation/prplMesh/pull/537#discussion_r355436664

File: common/beerocks/scripts/beerocks_utils.sh (hunk 4 above)
Comment (morantr) [reply common/beerocks/scripts/beerocks_utils.sh]: Ok, but in a different task.

File: framework/CMakeLists.txt (hunk 5 above)
Comment (morantr) [reply framework/CMakeLists.txt]: dropped commit.

File: controller/src/beerocks/master/tasks/rdkb/rdkb_wlan_task.cpp (hunk 6 above)
Comment (morantr) [reply controller/src/beerocks/master/tasks/rdkb/rdkb_wlan_task.cpp:272]: changed to your suggested commit message.

File: common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp (hunk 7 above)
Comment (morantr) [reply common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:581]: I don't know exactly. All I know that the station did not get the dissasoc frame when "tx=0" was in the hostapd command.

File: common/beerocks/bcl/include/bcl/son/son_wireless_utils.h (hunk 8 above)
Comment (morantr) [reply common/beerocks/bcl/include/bcl/son/son_wireless_utils.h:150]: I have no idea how this was changed. Anyway, it is not a big deal.

File: common/beerocks/bcl/include/bcl/son/son_wireless_utils.h (hunk 11 above)
Comment (morantr) [reply common/beerocks/bcl/include/bcl/son/son_wireless_utils.h]: removed spaces.

File: common/beerocks/bcl/source/son/son_wireless_utils.cpp (hunk 12 above)
Comment (morantr) [reply common/beerocks/bcl/source/son/son_wireless_utils.cpp]: done

File: common/beerocks/bcl/source/son/son_wireless_utils.cpp (hunk 13 above)
Comment (morantr) [reply common/beerocks/bcl/source/son/son_wireless_utils.cpp]: done

File: common/beerocks/bcl/source/son/son_wireless_utils.cpp (hunk 14 above)
Comment (morantr) [reply common/beerocks/bcl/source/son/son_wireless_utils.cpp]: Removed the __FUNCTION__.
I agree about the pipes. But the guy who did it said that it looks better in the log file.

File: common/beerocks/bcl/source/son/son_wireless_utils.cpp (hunk 19 above)
Comment (morantr) [reply common/beerocks/bcl/source/son/son_wireless_utils.cpp]: done

File: common/beerocks/bcl/source/son/son_wireless_utils.cpp (hunk 18 above)
Comment (morantr) [reply common/beerocks/bcl/source/son/son_wireless_utils.cpp]: done

File: common/beerocks/bcl/source/son/son_wireless_utils.cpp (hunk 15 above)
Comment (morantr) [reply common/beerocks/bcl/source/son/son_wireless_utils.cpp:78]: I removed the minimal_estimation as you suggested. So I'm leaving it as is.

File: common/beerocks/bcl/source/son/son_wireless_utils.cpp (hunk 16 above)
Comment (morantr) [reply common/beerocks/bcl/source/son/son_wireless_utils.cpp:104]: how it will make it shorter? We will have to write `value.gi_long_rate` and  `value.gi_short_rate` instead of `gi_long_rate` and `gi_short_rate`.
Anyway, it doesn't seem critical. I leaving it as is.

File: common/beerocks/bcl/source/son/son_wireless_utils.cpp (hunk 17 above)
Comment (morantr) [reply common/beerocks/bcl/source/son/son_wireless_utils.cpp]: It is the same result. I found it stupid to add 0.5 and then cast it to `int` so the number will be rounded up.

File: controller/src/beerocks/master/tasks/optimal_path_task.cpp (hunk 20 above)
Comment (morantr) [reply controller/src/beerocks/master/tasks/optimal_path_task.cpp]: done

File: common/beerocks/bcl/source/son/son_wireless_utils.cpp (hunk 17 above)
Comment (arnout) [reply common/beerocks/bcl/source/son/son_wireless_utils.cpp]: The original code was `int(float(est) / 10.0 + 0.5)`. Sinds int() rounds down, this rounds down if the last decimal is smaller than 5 and up if it is larger than 5. The new code is `int(ceil(est / 10.0))` which always rounds up.

See this test code:

```
#include <iostream>
#include <cmath>

int main() {
  int est = 14;
  std::cout << "div: " << (float(est) / 10.0) << " int: " << int(float(est)/10.0 + 0.5) << std::endl;
  std::cout << "div: " << (est / 10.0) << " ceil: " << ceil(est/10.0) << " int: " << int(ceil(est/10.0)) << std::endl;
}
```

output:

```
div: 1.4 int: 1
div: 1.4 ceil: 2 int: 2
```

Comment (morantr) [reply common/beerocks/bcl/source/son/son_wireless_utils.cpp]: I thought it will round to the nearest integer value.
Anyway, according to the comments. the one who coded it meant to round it up. So in that case, it fixes it.

File: common/beerocks/bcl/source/son/son_wireless_utils.cpp [hunk 21]
```diff
@@ -56,22 +50,27 @@ wireless_utils::estimate_ul_params(int ul_rssi, uint16_t sta_phy_tx_rate_100kb,
...
     if (sta_phy_tx_rate_100kb < phy_rate_table[0][0].bw_values[0].gi_long_rate) {
-        LOG(DEBUG) << __FUNCTION__ << " | Can not estimate UL parameters (STA phyrate is too low)";
-        return estimation_minimal;
+        LOG(DEBUG) << "Can not estimate UL parameters (STA phyrate is too low)";
+        estimation.rssi =
+            is_5ghz ? phy_rate_table[0][0].tx_power_5 : phy_rate_table[0][0].tx_power_2_4;
+        estimation.tx_power = int(ceil(phy_rate_table[0][0].bw_values[0].rssi / 10.0));
+        estimation.status   = ESTIMATION_FAILURE_BELOW_RANGE;
+
+        return estimation;
```
Comment (arnout) [inline common/beerocks/bcl/source/son/son_wireless_utils.cpp]: Yes, I like this a lot better.

Well, except that rssi and tx_power are swapped :-) Not so good...

File: agent/src/beerocks/monitor/beerocks_monitor_main.cpp (hunk 2 above)
Comment (vitalybu) [reply agent/src/beerocks/monitor/beerocks_monitor_main.cpp]: I think that in RDKB there's some sort of centralized logging cleanup service that expects any log files to have specific extension... I think we should keep it for RDKB only.

**Review by vitalybu — COMMENTED** [review]
That's not correct. The ARP monitor is used for discovering Ethernet clients (IP address is discovered by the DHCP monitor). AFAIK EasyMesh doesn't handle Ethernet clients, but we already have this built-in...

File: agent/config/beerocks_agent.conf.in [hunk 22]
```diff
@@ -13,7 +13,7 @@ vendor=Intel
 model=prplMesh
 ucc_listener_port=8002 # 0 - disabled
 ucc_listener_slave_hostap_iface=wlan0
-enable_arp_monitor=1
+enable_arp_monitor=0
```
Comment (vitalybu) [inline agent/config/beerocks_agent.conf.in:16]: That's not correct. The ARP monitor is used for discovering Ethernet clients (IP address is discovered by the DHCP monitor). AFAIK EasyMesh doesn't handle Ethernet clients, but we already have this built-in...

File: agent/src/beerocks/monitor/monitor_stats.cpp [hunk 23]
```diff
@@ -185,6 +185,152 @@ void monitor_stats::process()
...
+            << "vap client_rx_load_tot_curr > 100, changing to radio client_rx_load_tot_curr = 100 "
+            << std::endl;
+        radio_stats.client_rx_load_tot_curr = 100;
+    }
+
+    radio_stats.channel_load_idle = 100 - radio_stats.channel_load_tot_curr;
+    radio_stats.channel_load_others =
+        radio_stats.channel_load_tot_curr -
+        (radio_stats.client_tx_load_tot_curr + radio_stats.client_rx_load_tot_curr);
+    //LOG(DEBUG) << "STATS_MEASUREMENT_vap: "<< vap_node;
```
Comment (vitalybu) [inline agent/src/beerocks/monitor/monitor_stats.cpp]: Remove commented prints

File: common/beerocks/bcl/source/beerocks_os_utils.cpp [hunk 24]
```diff
@@ -86,10 +87,15 @@ void os_utils::kill_pid(std::string path, std::string file_name)
 {
     int pid_out;
     if (is_pid_running(path, file_name, &pid_out)) {
-        std::cout << "kill_pid() kill pid=" << pid_out << std::endl;
+        LOG(DEBUG) << __FUNCTION__ << " SIGTERM pid=" << pid_out << std::endl;
         kill(pid_out, SIGTERM);
-        sleep(2); //TODO: find alternative solution
+        auto timeout = std::chrono::steady_clock::now() + std::chrono::seconds(15);
```
Comment (vitalybu) [inline common/beerocks/bcl/source/beerocks_os_utils.cpp:92]: Timeout value should be exported as an argument.
But overall, busy-wait is never a good idea.
Maybe sleep for short intervals inside the while loop?
(NOTE: `waitpid()` doesn't support timeouts [without ugly alarm/select workaround])

File: agent/src/beerocks/monitor/beerocks_monitor_main.cpp (hunk 2 above)
Comment (tomereli) [reply agent/src/beerocks/monitor/beerocks_monitor_main.cpp]: I have no idea. I remember Yaniv did some hacks for rdkb logging and I don't think we should take them as is. Please send him a mail with all the review questions.

File: common/beerocks/bcl/source/beerocks_logging.cpp (hunk 3 above)
Comment (tomereli) [reply common/beerocks/bcl/source/beerocks_logging.cpp]: Should be in the mail to Yaniv.

File: common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp (hunk 7 above)
Comment (tomereli) [reply common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:581]: Not knowing is fine, but putting such changes to the code without knowing and understanding the change is not. Please ask Alex, Yaniv, and whoever was involved with the changes in rdkb (hint - they sit next to you) about this and any other question you didn't know the answer to.
Then add a comment in the code and update the commit message.

File: common/beerocks/scripts/beerocks_utils.sh (hunk 4 above)
Comment (tomereli) [reply common/beerocks/scripts/beerocks_utils.sh]: So please open a new task for it.

File: agent/config/beerocks_agent.conf.in (hunk 22 above)
Comment (morantr) [reply agent/config/beerocks_agent.conf.in:16]: In the commit message, I wrote that it observed to cause high CPU cycles and that this is a bandaid solution and better solution must be found. I also added a fixup that disable it only on RDKB.

**Comment by morantr** [comment]
> So move it to the BPL and let each platform dB decide

Not as part of this PR. There is already a task for moving configuration to the database.

File: agent/src/beerocks/monitor/monitor_stats.cpp (hunk 23 above)
Comment (morantr) [reply agent/src/beerocks/monitor/monitor_stats.cpp]: done

**Review by arnout — APPROVED** [review]
Except for the rename of the log file, all review feedback was taken into account. So I rebased on master and split the log file commit. Only the specific directory on rdkb is retained; the file renaming is on a separate commit that is pushed to branch dev/change-log-file-names.
